docs/cli_development.md

```diff
@@ -0,0 +1,177 @@
+# QMK CLI Development
+
+This document has useful information for developers wishing to write new `qmk` subcommands.
+
+# Overview
+
+The QMK CLI operates using the subcommand pattern made famous by git. The main `qmk` script is simply there to setup the environment and pick the correct entrypoint to run.
+
+There are actually 2 different `qmk` programs, but they operate in very similar ways. There is an installable `qmk`, also referred to as the global CLI, which can operate on many different `qmk_firmware` repositories. There is also the bundled `bin/qmk`, also referred to the local CLI, which always operates on the `qmk_firmware` it is located in. In both cases the `qmk` script prepares the environment, loads all of the subcommand modules, and then executes the entrypoint designated by the user's CLI command.
+
+Each subcommand is a self-contained module with an entrypoint (decorated by `@cli.subcommand()`) that performs some action and returns a shell returncode, or None.
+
+# Subcommands
+
+[MILC](https://github.com/clueboard/milc) is the CLI framework `qmk` uses to handle argument parsing, configuration, logging, and many other features. It lets you focus on writing your tool without wasting your time writing glue code.
+
+Subcommands in the local CLI are always found in `qmk_firmware/lib/python/qmk/cli`. Subcommands in the global CLI are always found in `qmk_cli/qmk_cli/subcommands`. Most subcommands should be located in the local CLI. The global CLI subcommands should be limited to those few needed to setup a `qmk_firmware` environment.
```

Comment:
the global vs local thing is a bit confusing. Which one should developer prioritize? what kinds of things should be local? Should it be moved to an appendix?